Set up linking for web

zecakeh opened the pull request:
So that the web client has nice urls like this:
- home: `/:colUid?`,
- CollectionEdit: `/:colUid/edit`,
- CollectionChangelog: `/:colUid/manage`,
- CollectionMembers: `/:colUid/members`,
- ItemEdit: `/:colUid/:itemUid`,
- all the others have their name or title in lowercase, except About and DebugLogs thaht have the `settings/` prefix too.

I defined arbitrary prefixes for deep linking because TypeScript requested it, but I didn't set it up for native apps.

I had to change a bit the `Drawer` navigation so `colUid` is set as an empty string instead of undefined otherwise it shows up it the url. So I did a little change in `NoteListScreen` that's pretty useless I guess to avoid querying the `cacheCollection` with an empty string.

This fixes #27. What's requested in the title anyway. We still need to display a pointer when we hover components that trigger navigation. If I remember correctly that was being worked on both in `react-native-paper` and `react-navigation`.

Tested on Firefox Linux and Android

tasn commented:
As the one who opened the original ticket this makes me so happy! Thanks a lot for this!

I found a bug though: after I login, and the page switches to "home", the path is still "login". I guess it's a bug elsewhere in the code, so don't worry about it if hard to fix, but would be nice.

As for the pointer showing that you mentioned: it's also a bug in the libraries you mentioned (potentially), but it's also something we are doing wrong. We are currently triggering the page transitions in js (in the callbacks), rather than using react-navigation links. I think that if we use them that will work. The pointer thing that react-native-paper isn't showing for buttons is something that will fix the pointer icon for buttons, but it won't fix e.g. the link showing in the browser when hovering over a button.

zecakeh commented:
For the pointers and links it depends, sometimes it can work natively: [App Drawer](https://snack.expo.io/?platform=android&name=Getting%20started%20%7C%20React%20Navigation&dependencies=%40expo%2Fvector-icons%40*%2C%40react-native-community%2Fmasked-view%40*%2C%40react-navigation%2Fbottom-tabs%40%5E5.8.0%2C%40react-navigation%2Fdrawer%40%5E5.9.0%2C%40react-navigation%2Fmaterial-bottom-tabs%40%5E5.2.16%2C%40react-navigation%2Fmaterial-top-tabs%40%5E5.2.16%2C%40react-navigation%2Fnative%40%5E5.7.3%2C%40react-navigation%2Fstack%40%5E5.9.0%2Creact-native-paper%40%5E4.0.1%2Creact-native-reanimated%40*%2Creact-native-safe-area-context%40*%2Creact-native-gesture-handler%40*%2Creact-native-screens%40*%2Creact-native-tab-view%40%5E2.15.1&sourceUrl=https%3A%2F%2Freactnavigation.org%2Fexamples%2F5.x%2Fdrawer-open-close-toggle.js)

Although I guess react-native-paper might break that.

But the rest of the time we should use [`useLinkProps`](https://reactnavigation.org/docs/use-link-props) to create components only for web.

I'm gonna look at the login thing, it might be my fault.

tasn commented:
I think there's maybe even a Link widget in react-navigation (I may be misremembering), though anyhow, yeah, the solution is still the same, we need to use a Link widget whether we have to implement it or not.

zecakeh commented:
Ok so actually the login in the url is because of my linking config. Since the home screen is at `/` and it can receive a `colUid` param, it treats `login` as such.

So a solution would be to have the home and notebooks at `/notes`, or have the home at `/` and the notebooks at `/notes/colUid` or something.

That makes me think that we don't have a 404 page, and we don't do anything if `colUid` and/or `itemUid` is wrong. If a url like `/colUid/ItemUid` is wrong, that makes the app crash in debug.

tasn commented:
I'm happy with having everything under `/notes` as long as `/` redirects. Also happy with your second suggestion.

404: yeah, it's because we didn't have any links until now, so I haven't bothered with that. :P

zecakeh commented:
- Made the URL changes (everything is under `/notes`)
- Added a 404 page
- Redirected the screens that use `colUid` and `itemUid` in the route params to 404 with a custom message if they are not valid
- Rebased on master

tasn commented on src/screens/NoteListScreen.tsx:
```diff
@@ -80,10 +80,17 @@ export default function NoteListScreen(props: PropsType) {
+  const cacheCollection = (colUid && colUid.length > 0) ? cacheCollections.get(colUid) : undefined;
+  
+  if (colUid && colUid.length > 0 && cacheCollection == null) {
+    navigation.navigate("404", {
```
Maybe have a `navigateTo404(navigation, message="This notebook can't be found",  title="Not found")` utility function? So in most cases we can just call `navigateTo404(navigation)` and it'll work nicely? Having the text spread around everywhere feels very redundant.

zecakeh replied:
Sure, we can do that! I just thought at first the message would change more but it's mainly colUid in the params…
